Content.Client/Access/UI/AgentIDCardWindow.xaml.cs

```diff
@@ -1,24 +1,96 @@
-using Robust.Client.UserInterface.CustomControls;
+using Content.Client.Stylesheets;
+using Content.Shared.StatusIcon;
 using Robust.Client.AutoGenerated;
+using Robust.Client.GameObjects;
+using Robust.Client.UserInterface;
+using Robust.Client.UserInterface.Controls;
+using Robust.Client.UserInterface.CustomControls;
 using Robust.Client.UserInterface.XAML;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Utility;
+using System.Linq;
+using System.Numerics;
 
 namespace Content.Client.Access.UI
 {
     [GenerateTypedNameReferences]
     public sealed partial class AgentIDCardWindow : DefaultWindow
     {
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly IEntitySystemManager _entitySystem = default!;
+        private readonly SpriteSystem _spriteSystem = default!;
+
+        private const string JobIconPrarentId = "JobIcon";
+        private const int JobIconColumnCount = 10;
+
         public event Action<string>? OnNameChanged;
         public event Action<string>? OnJobChanged;
+        public readonly List<Button> JobIconButtons = new();
+        public readonly Dictionary<int, string> StateByIndex = new();
 
         public AgentIDCardWindow()
         {
             RobustXamlLoader.Load(this);
+            IoCManager.InjectDependencies(this);
+            _spriteSystem = _entitySystem.GetEntitySystem<SpriteSystem>();
 
             NameLineEdit.OnTextEntered += e => OnNameChanged?.Invoke(e.Text);
             NameLineEdit.OnFocusExit += e => OnNameChanged?.Invoke(e.Text);
 
             JobLineEdit.OnTextEntered += e => OnJobChanged?.Invoke(e.Text);
             JobLineEdit.OnFocusExit += e => OnJobChanged?.Invoke(e.Text);
+
+            var jobIconGroup = new ButtonGroup();
+            var allJobIcons = _prototypeManager.EnumeratePrototypes<StatusIconPrototype>();
+            var i = 0;
+            foreach (var jobIcon in allJobIcons)
+            {
+                if (!(jobIcon.Parents?.Contains(JobIconPrarentId) ?? false))
+                {
+                    continue;
+                }
```

Comment:
You should almost never check against a prototype's parents. It is not meant to be used to store data about a prototype and is only meant to be used for convenience in yaml. Doing this  can also introduce unexpected bugs in future. E.g. if someone ever moves all security icons to their own RSI and adds a new parent (i.e., `JobIcon` -> `JobIconSecurity` -> `JobIconSecurityOfficer`) then this would break as `JobIcon` is no longer a direct parent of `JobIconSecurityOfficer`.

There are two decent ways I can think of to add icon selection restrictions:
- Add a new data-field to `JobPrototype`, either a bool or some icon-category enum, then set it to some value in the `JobIcon` parent & check the value here.
- Add a new data-field to the agent ID card that has a set of all allowed IconPrototypes, then enumerate over that set instead of over all prototypes

Reply:
I don't think the example works. Parents is an array, I would expect the parent of my parent to be in the array.

Reply:
Reading the code in PrototypeManager this seems to be the most pointless field I've ever seen, nevermind.

Reply:
The array is for composition, it allows you to inherit from multiple parents simultaneously, but its an understandable source of confusion.